### PR TITLE
(PUP-10857) Find groups locally with useradd provider

### PR DESCRIPTION
**Pull request by luchihoratiu**
Before this commit, `forcelocal` had no effect on the `groups` field from the `useradd` provider. This change allows to get groups from `/etc/groups` when `forcelocal` is set to `true`.

This PR also adds `forcelocal` option effect on user gid from the `useradd` provider and fixes the retrieval of arrays from the property list.

### CODE REVIEW

File: lib/puppet/provider/group/groupadd.rb [hunk 1]
```diff
@@ -130,16 +130,21 @@ def purge_members
...
+
+      @groups = []
+      Puppet::FileSystem.each_line(group_file) do |line|
+        group = line.chomp.split(':')
+        @groups << Hash[group_keys.zip(group)]
       end
     end
-    false
+
+    @groups.select { |param| param[key] == value }.first || false
```
Comment (gimmyxd) [inline lib/puppet/provider/group/groupadd.rb:148]: @groups.find { |param| param[key] == value } is faster than `.select.first`

File: lib/puppet/property/list.rb [hunk 2]
```diff
@@ -47,7 +47,7 @@ def retrieve
         #ok, some 'convention' if the list property is named groups, provider should implement a groups method
         tmp = provider.send(name) if provider
         if tmp && tmp != :absent
-          return tmp.split(delimiter)
+          return Array === tmp ? tmp : tmp.split(delimiter)
```
Comment (gimmyxd) [inline lib/puppet/property/list.rb:50]: Usually one would use `.instance_of?` for checking that `tmp` is strictly an object of Array type. 
`===` returns true also if `tmp` is an object of a subclass of Array.

**Review by gimmyxd — APPROVED** [review]

**Comment by shaigy** [comment]
@luchihoratiu @mihaibuzgau @GabrielNagy This is causing a corrective changeduring every puppet run for the user pe-bolt-server defined with groups [ 'pe-bolt-server', 'pe-orchestration-services'] here https://github.com/puppetlabs/puppet-enterprise-modules/blob/35ccf3619975a5ccf2518fbb79f91fdb738b60bf/modules/puppet_enterprise/manifests/profile/bolt_server.pp#L89-#L91 causing test failures in PE CI

**Comment by luchihoratiu** [comment]
@shaigy #8504 should solve the issue

**Pull request merged**